components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/threatprotection/analyzer/XMLAnalyzer.java

```diff
@@ -0,0 +1,117 @@
+import javax.xml.stream.XMLStreamException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+
+
+/**
+ * Implementation of APIMThreatAnalyzer for XML Payloads
+ */
+public class XMLAnalyzer implements APIMThreatAnalyzer {
+
+    private Logger log = LoggerFactory.getLogger(XMLAnalyzer.class);
+    private static final String XML_THREAT_PROTECTION_MSG_PREFIX = "Threat Protection-XML: ";
+    private XMLInputFactory factory;
+    private boolean enabled = true;
+
+    public XMLAnalyzer() {
+        factory = WstxInputFactory.newInstance();
+    }
+
+    /**
+     * Create a XMLAnalyzer using default configuration values
+     */
+    public void configure(XMLConfig config) {
+        boolean dtdEnabled = config.isDtdEnabled();
+        boolean externalEntitiesEnabled = config.isExternalEntitiesEnabled();
+        Integer maxDepth = config.getMaxDepth();
+        Integer maxElementCount = config.getMaxElementCount();
+        Integer maxAttributeCount = config.getMaxAttributeCount();
+        Integer maxAttributeLength = config.getMaxAttributeLength();
+        Integer entityExpansionLimit = config.getEntityExpansionLimit();
+        Integer maxChildrenPerElement = config.getMaxChildrenPerElement();
+        factory.setProperty(XMLInputFactory.SUPPORT_DTD, dtdEnabled);
+        factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, externalEntitiesEnabled);
+        factory.setProperty(ThreatProtectorConstants.P_MAX_ATTRIBUTE_SIZE, maxAttributeLength);
+        factory.setProperty(ThreatProtectorConstants.P_MAX_ATTRIBUTES_PER_ELEMENT, maxAttributeCount);
+        factory.setProperty(ThreatProtectorConstants.P_MAX_ELEMENT_DEPTH, maxDepth);
+        factory.setProperty(ThreatProtectorConstants.P_MAX_CHILDREN_PER_ELEMENT, maxChildrenPerElement);
+        factory.setProperty(ThreatProtectorConstants.P_MAX_ENTITY_COUNT, entityExpansionLimit);
+        factory.setProperty(ThreatProtectorConstants.P_MAX_ELEMENT_COUNT, maxElementCount);
+    }
+
+    @Override
+    public void configure(JSONConfig config) {
+        throw new UnsupportedOperationException("This method is not supported on this instance");
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    /**
+     * @param in xml payload
+     * @throws APIMThreatAnalyzerException
+     */
+    @Override
+    public void  analyze(InputStream in, String apiContext) throws APIMThreatAnalyzerException {
+        Reader reader = null;
+        XMLEventReader xmlEventReaderReader = null;
+        try {
+            reader = new InputStreamReader(in);
+            xmlEventReaderReader = factory.createXMLEventReader(reader);
+            while (xmlEventReaderReader.hasNext()) {
+                xmlEventReaderReader.nextEvent();
+            }
+        } catch (XMLStreamException e) {
+            throw new APIMThreatAnalyzerException("XML Validation Failed: due to "+ e.getMessage());
```

Comment:
If we remove e from the exception. We are missing the stack trace of the error, which will make it hard to debug

Reply:
fixed the issue.